File: sdks/node/test/core/wireVectors.test.ts
// Asserts the shared cross-SDK wire vectors.
//
// `contracts/wire-vectors.json` pins the bytes of the CBOR call envelope. Every
// SDK runs this same file against its own serializer, so an encoding change
// fails the runtime that made it instead of quietly producing payloads its peers
// cannot read.

import { existsSync, readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { CborSerializer } from "../../src/serializers";

interface EncodeCase {
  name: string;
  args: unknown[];
  kwargs: Record<string, unknown>;
  hex: string;
}

interface DecodeCase {
  name: string;
  hex: string;
  args?: unknown[];
  kwargs?: Record<string, unknown>;
  round_trip_only?: boolean;
}

/** Walk up to the repository root rather than counting directories. */
function repoRoot(): string {
  let dir = dirname(fileURLToPath(import.meta.url));
  for (;;) {
    if (existsSync(join(dir, "contracts", "wire-vectors.json"))) {
      return dir;
    }
    const parent = resolve(dir, "..");
    if (parent === dir) {
      throw new Error("contracts/wire-vectors.json not found above this test");
    }
    dir = parent;
  }
}

const ROOT = repoRoot();
const vectors: { encode: EncodeCase[]; decode_only: DecodeCase[] } = JSON.parse(
  readFileSync(join(ROOT, "contracts", "wire-vectors.json"), "utf8"),
);
const BINDING_CONTRACT = join(ROOT, "crates", "taskito-core", "BINDING_CONTRACT.md");

// A run of space-separated hex byte pairs inside backticks, as the doc writes them.
const DOCUMENTED_BYTES = /`((?:[0-9a-f]{2} )+[0-9a-f]{2})`/;

const hex = (bytes: Uint8Array) => Buffer.from(bytes).toString("hex");

/**
 * JavaScript has no keyword arguments, so `serializeCall` always writes an empty
 * kwargs map and cannot produce these cases. Decoding them is still required —
 * a producer in a runtime that does have keyword arguments can enqueue one.
 */
const encodable = vectors.encode.filter((c) => Object.keys(c.kwargs).length === 0);

describe("cross-SDK wire vectors", () => {
  it.each(
    encodable.map((c) => [c.name, c] as const),
  )("encodes %s to the pinned bytes", (_name, testCase) => {
    expect(hex(new CborSerializer().serializeCall(testCase.args))).toBe(testCase.hex);
  });

  it.each(
    vectors.encode.map((c) => [c.name, c] as const),
  )("decodes %s from the pinned bytes", (_name, testCase) => {
    const decoded = new CborSerializer().deserialize(Buffer.from(testCase.hex, "hex"));
    expect(decoded).toEqual([testCase.args, testCase.kwargs]);
  });

  it.each(vectors.decode_only.map((c) => [c.name, c] as const))("decodes %s", (_name, testCase) => {
    const serializer = new CborSerializer();
    const raw = Buffer.from(testCase.hex, "hex");
    const [args, kwargs] = serializer.deserialize(raw) as [unknown[], Record<string, unknown>];

    if (testCase.round_trip_only) {
      // The value has no lossless JSON form, so re-encoding is the assertion.
      expect(kwargs).toEqual({});
      expect(hex(serializer.serializeCall(args))).toBe(testCase.hex);
    } else {
      expect(args).toEqual(testCase.args);
      expect(kwargs).toEqual(testCase.kwargs);
    }
  });

  // BINDING_CONTRACT.md restates the call vector; drift there is a silent contract change.
  it("keeps the vector quoted in BINDING_CONTRACT.md", () => {
    const quoted = readFileSync(BINDING_CONTRACT, "utf8")
      .split("\n")
      .find((line) => line.includes('call `f(1, "a")`'));
    if (quoted === undefined) {
      throw new Error(`no documented call vector in ${BINDING_CONTRACT}`);
    }

    const documented = DOCUMENTED_BYTES.exec(quoted)?.[1];
    if (documented === undefined) {
      throw new Error(`no hex byte run in: ${quoted}`);
    }

    const shared = vectors.encode.find((c) => c.name === "contract-vector");
    expect(documented.replaceAll(" ", "")).toBe(shared?.hex);
  });
});

File: sdks/java/src/test/java/org/byteveda/taskito/serialization/WireVectorsTest.java
package org.byteveda.taskito.serialization;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

/**
 * Asserts the shared cross-SDK wire vectors.
 *
 * <p>{@code contracts/wire-vectors.json} pins the bytes of the CBOR call
 * envelope. Every SDK runs this same file against its own serializer, so an
 * encoding change fails the runtime that made it instead of quietly producing
 * payloads its peers cannot read.
 */
class WireVectorsTest {
    private static final ObjectMapper JSON = new ObjectMapper();
    private static final JsonNode VECTORS = load();

    private final CborSerializer cbor = new CborSerializer();

    /** Walk up to the repository root rather than counting directories. */
    private static JsonNode load() {
        Path dir = Paths.get("").toAbsolutePath();
        while (dir != null) {
            Path candidate = dir.resolve("contracts").resolve("wire-vectors.json");
            if (Files.isRegularFile(candidate)) {
                try {
                    return JSON.readTree(candidate.toFile());
                } catch (IOException e) {
                    throw new IllegalStateException("failed to read " + candidate, e);
                }
            }
            dir = dir.getParent();
        }
        throw new IllegalStateException("contracts/wire-vectors.json not found above the working directory");
    }

    private static List<JsonNode> cases(String section) {
        List<JsonNode> out = new ArrayList<>();
        VECTORS.get(section).forEach(out::add);
        return out;
    }

    /**
     * Java has no keyword arguments and {@code enqueue} takes a single payload,
     * so the serializer can only produce a one-positional-argument call with an
     * empty kwargs map. Every other case is still decoded below — a producer in
     * another runtime can enqueue one.
     */
    static Stream<JsonNode> encodable() {
        return cases("encode").stream()
                .filter(c -> c.get("kwargs").isEmpty())
                .filter(c -> c.get("args").size() == 1);
    }

    static Stream<JsonNode> encodeCases() {
        return cases("encode").stream();
    }

    static Stream<JsonNode> decodeOnlyCases() {
        return cases("decode_only").stream();
    }

    private static String hex(byte[] bytes) {
        StringBuilder out = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            out.append(Character.forDigit((b >> 4) & 0xf, 16)).append(Character.forDigit(b & 0xf, 16));
        }
        return out.toString();
    }

    private static byte[] unhex(String text) {
        byte[] out = new byte[text.length() / 2];
        for (int i = 0; i < out.length; i++) {
            out[i] = (byte) Integer.parseInt(text.substring(i * 2, i * 2 + 2), 16);
        }
        return out;
    }

    @ParameterizedTest(name = "encodes {0}")
    @MethodSource("encodable")
    void encodesToThePinnedBytes(JsonNode testCase) {
        Object payload = JSON.convertValue(testCase.get("args").get(0), Object.class);

        assertEquals(testCase.get("hex").asText(), hex(cbor.serializeCall(payload)));
    }

    /**
     * Guards the encoding choice that makes those bytes match in the first place.
     *
     * <p>Jackson streams values, so left alone it writes CBOR maps with indefinite
     * length ({@code 0xbf ... 0xff}) where this contract pins definite length
     * ({@code 0xa0}). Both decode identically, so a regression here would not break
     * interoperability — it would silently break the cross-SDK guarantee documented
     * on {@code IdempotencyKeys}, because the automatic {@code auto:} key hashes the
     * serialized payload. Every call ends in the empty kwargs map, so a regression
     * would shift <em>every</em> payload's key at once.
     */
    @Test
    void writesDefiniteLengthContainers() {
        String encoded = hex(cbor.serializeCall("x"));

        assertEquals("0282816178a0", encoded, "kwargs must be a definite-length empty map (a0), not bf..ff");
        assertTrue(!encoded.contains("bf"), "no indefinite-length container header may appear: " + encoded);
    }

    /** Payloads written before the definite-length fix must still decode. */
    @Test
    void decodesLegacyIndefiniteLengthPayloads() {
        // 02 82 81 a1 6161 6162 bfff — args definite, kwargs the old indefinite empty map.
        Object call = cbor.deserialize(unhex("028281a161616162bfff"), Object.class);

        assertEquals("[[{a=b}], {}]", String.valueOf(call));
    }

    @ParameterizedTest(name = "decodes {0}")
    @MethodSource("encodeCases")
    void decodesThePinnedBytes(JsonNode testCase) {
        byte[] raw = unhex(testCase.get("hex").asText());
        JsonNode call = JSON.valueToTree(cbor.deserialize(raw, Object.class));

        assertEquals(testCase.get("args"), call.get(0));
        assertEquals(testCase.get("kwargs"), call.get(1));
    }

    @ParameterizedTest(name = "decodes {0}")
    @MethodSource("decodeOnlyCases")
    void decodeOnlyVectors(JsonNode testCase) {
        byte[] raw = unhex(testCase.get("hex").asText());
        Object call = cbor.deserialize(raw, Object.class);

        if (testCase.path("round_trip_only").asBoolean()) {
            // The value has no lossless JSON form, so re-encoding is the assertion.
            List<?> parts = (List<?>) call;
            assertTrue(((java.util.Map<?, ?>) parts.get(1)).isEmpty());
            Object payload = ((List<?>) parts.get(0)).get(0);

            assertEquals(testCase.get("hex").asText(), hex(cbor.serializeCall(payload)));
        } else {
            JsonNode decoded = JSON.valueToTree(call);
            assertEquals(testCase.get("args"), decoded.get(0));
            assertEquals(testCase.get("kwargs"), decoded.get(1));
        }
    }

    @Test
    void keepsTheVectorQuotedInTheBindingContract() {
        JsonNode documented = cases("encode").stream()
                .filter(c -> "contract-vector".equals(c.get("name").asText()))
                .findFirst()
                .orElseThrow();
        assertEquals("028282016161a0", documented.get("hex").asText());
    }
}
